### ISSUE TITLE
Match + Gym + Deliver doesn't use correct provisioning profile.

### ISSUE THREAD

**Issue by jcampbell05** [fastlane/fastlane#4694]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description

When running a lane which invokes match, builds via gym and then uses deliver. I get an error from iTunes Connect that the IPA was uploaded with a development provisioning profile. I opened up the IPA and sure enougth it was.

On inspection I discovered my default certificate type in my `matchfile` was `development` so I change that to `appstore`. This then updated the SIGH certificate type to `app-store`.

However I still got the same error. I then made match only download the certificates for the appstore, incase gym was pulling in the incorrect type.

Again even though the SIGH certificate type was for the `appstore` and the matchfile species `appstore` as the default, gym and deliver seems to be building a IPA with a development provisioning profile.

This is especially odd considering gym is building a release scheme.

##### Complete output when running fastlane, including the stack trace and command used

Command ran `fastlane deploy`

[Log.txt](https://github.com/fastlane/fastlane/files/269492/Log.txt)

### Configuration Files

Please copy the complete content of your `Fastfile` and any other configuration files you use below:

Fastfile:
```ruby
# Customise this file, documentation can be found here:
# https://github.com/KrauseFx/fastlane/tree/master/docs
# All available actions: https://github.com/KrauseFx/fastlane/blob/master/docs/Actions.md
# can also be listed using the `fastlane actions` command

# Change the syntax highlighting to Ruby
# All lines starting with a # are ignored when running `fastlane`

# By default, fastlane will send which actions are used
# No personal data is shared, more information on https://github.com/fastlane/enhancer
# Uncomment the following line to opt out
# opt_out_usage

# If you want to automatically update fastlane if a new version is available:
# update_fastlane

# This is the minimum version number required.
# Update this, if you use features of a newer version
fastlane_version "1.89.0"

####### TODO #######

# - Check user is in correct branch
# - Implement Git workflow into Lanes (i.e tagging and generating release branch)
# - Reduce length of change log

####### Settings #######

APP_IDENTIFIER = "com.forever-beta.sup"
CRASHLYTICS_API_KEY = <redacted>
CRASHLYTICS_BUILD_SECRET = <redacted>
CRASHLYTICS_PATH = "./Pods/Crashlytics/Crashlytics.framework"
IPA_PATH = "./Sup.ipa"
SLACK_URL = <redacted>

####### Methods #######

def inform_team(message)
  return if Helper.is_ci?

  slack(
    message: message,
    default_payloads: [],
    channel: "#ios"
  )
end

def setup_enviroment_variables
  ENV["SLACK_URL"] = SLACK_URL

  # Turn on workaround for https://github.com/CocoaPods/CocoaPods/issues/4178
  ENV["GYM_USE_GENERIC_ARCHIVE_FIX"] = "1"

  # Turn on workaround for https://github.com/CocoaPods/CocoaPods/issues/4178
  ENV["FASTLANE_EXPLICIT_OPEN_SIMULATOR"] = "1"
end

def setup_project
  inform_team("Generating Project And Installing Dependencies")
  xcake2
end

def install_fix_code_xcode_plugin
  # Install the FixCode plugin so that developers know we use match and not to
  # use `Fix Issue` inside of Xcode
  install_xcode_plugin(
    url: "https://github.com/fastlane/FixCode/releases/download/0.2.0/FixCode.xcplugin.zip"
  )
end

def download_certificate_and_profile(type, newDevices = false)
  match(app_identifier: APP_IDENTIFIER, type: type, force_for_new_devices: newDevices)
  match(app_identifier: "#{APP_IDENTIFIER}.watchkitapp", type: type, force_for_new_devices: newDevices)
  match(app_identifier: "#{APP_IDENTIFIER}.watchkitapp.watchkitextension", type: type, force_for_new_devices: newDevices)
end

def download_certificates_and_profiles(newDevices = false)
  ["development", "adhoc", "appstore"].each do |type|
    download_certificate_and_profile(type, newDevices)
  end
end

def bump_build_version
  build_number = number_of_commits
  increment_build_number(build_number: build_number)
end

def build_app(scheme, certificate_type)

  inform_team("Downloading Certificate And Provisoning Profile")
  download_certificate_and_profile(certificate_type)

  setup_project

  inform_team("Bumping Build Version")
  bump_build_version

  inform_team("Building App")
  gym(
    scheme: scheme,
    use_legacy_build_api: true
  )
end

def build_app_for_beta_testing(scheme, certificate_type)
  # Disabled until we can find a way to control where the changelog begins from
  # since it exceeds the lengthe crashyltics likes.
  # changelog_from_git_commits

  badge
  build_app(scheme, certificate_type)
end

def push_to_crashlytics(list)
  inform_team("Pushing to crashlytics")
  crashlytics(
    crashlytics_path: CRASHLYTICS_PATH,
    api_token: CRASHLYTICS_API_KEY,
    build_secret: CRASHLYTICS_BUILD_SECRET,
    ipa_path: IPA_PATH,
    groups: list
  )
  inform_team("Uploaded Build to Crashlytics")
end

def test_app
  setup_project
  scan(
    scheme: "Sup-Debug"
  )
end

####### Lanes #######

default_platform :ios

platform :ios do
  before_all do
    ensure_git_status_clean
    setup_enviroment_variables
  end

  desc "Runs all the tests"
  desc "This will do the following: "
  desc "- Generate the Xcode Project based on the Cakefile"
  desc "- Install CocoaPods"
  desc "- Make sure the profiles are up to date and download the latest one"
  desc "- Build and run the unit tests for the app."
  lane :test do
    test_app
  end

  desc "This is ran for the CI system, this is an alias for the `test` lane."
  desc "This will do the following: "
  desc "- Generate the Xcode Project based on the Cakefile"
  desc "- Install CocoaPods"
  desc "- Make sure the profiles are up to date and download the latest one"
  desc "- Build and run the unit tests for the app."
  lane :pr do
    test_app
  end

  desc "Generates a new Xcode Project"
  desc "This will do the following: "
  desc "- Generate the Xcode Project based on the Cakefile"
  desc "- Install CocoaPods"
  desc "- Make sure the profiles are up to date and download the latest one"
  desc "- Install the Fix Code Xcode Plugin."
  lane :xcode do
    setup_project
    download_certificates_and_profiles
    install_fix_code_xcode_plugin
  end

  desc "Build and upload a new nightly build to Crashlytics"
  desc "This will do the following: "
  desc "- Generate the Xcode Project based on the Cakefile"
  desc "- Install CocoaPods"
  desc "- Make sure the profiles are up to date and download the latest one"
  desc "- Add a badge to the app icon"
  desc "- Generate a changelog from the git commits"
  desc "- Build, sign and upload the app"
  desc "- Send an email to all nightly testers."
  lane :nightly do
    inform_team("Pushing Nightly")
    build_app_for_beta_testing("Sup-Staging", "adhoc")
    push_to_crashlytics("sup-nightly")
  end

  desc "Build and upload a new beta build to Crashlytics"
  desc "This will do the following: "
  desc "- Generate the Xcode Project based on the Cakefile"
  desc "- Install CocoaPods"
  desc "- Make sure the profiles are up to date and download the latest one"
  desc "- Add a badge to the app icon"
  desc "- Generate a changelog from the git commits"
  desc "- Build, sign and upload the app"
  desc "- Send an email to all internal testers."
  lane :internal_beta do
    inform_team("Pushing Internal Beta")
    build_app_for_beta_testing("Sup-Release", "appstore")
    push_to_crashlytics("sup")
  end

  desc "Build and upload a new build to Apple TestFlight"
  desc "This will do the following: "
  desc "- Generate the Xcode Project based on the Cakefile"
  desc "- Install CocoaPods"
  desc "- Make sure the profiles are up to date and download the latest one"
  desc "- Add a badge to the app icon"
  desc "- Generate a changelog from the git commits"
  desc "- Build, sign and upload the app"
  desc "- This will **not** send an email to all testers, it will only be uploaded to the new TestFlight."
  lane :external_beta do
    inform_team("Pushing External Beta")
    build_app_for_beta_testing("Sup-Release", "appstore")
    pilot
  end

  desc "Submit a new version to the App Store"
  desc "This will do the following: "
  desc "- Generate the Xcode Project based on the Cakefile"
  desc "- Install CocoaPods"
  desc "- Make sure the profiles are up to date and download the latest one"
  desc "- Do a build version bump and push it"
  desc "- Collect the app metadata from `./fastlane/metadata`"
  desc "- Upload screenshots + app metadata"
  desc "- Build, sign and upload the app"
  desc "- Upload dSYM files to Crashlytics"
  desc "This will **not** submit the app for review."
  lane :deploy do
    inform_team("Deploying To App Store")
    build_app("Sup-Release", "appstore")

    inform_team("Uploading to iTunes Connect")
    deliver
    inform_team("Uploaded Build #{Actions.lane_context[SharedValues::BUILD_NUMBER]} to iTunes Connect")

    inform_team("Uploading dSYMs to Crashlytics")
    upload_symbols_to_crashlytics
  end

  desc "Updates the dsyms uploaded to crashlytics from iTunes Connect"
  desc "This will do the following: "
  desc "- Download all dSYM files from iTunes Connect"
  desc "- Upload them to Crashlytics"
  desc "- Delete the local dSYM files"
  lane :refresh_dsyms do
     download_dsyms                  # Download dSYM files from iTunes Connect
     upload_symbols_to_crashlytics   # Upload them to Crashlytics
     clean_build_artifacts           # Delete the local dSYM files
  end

  desc "Make sure the profiles are up to date and download the latest one when we add new devices"
  lane :regenerate_for_new_devices do
    download_certificates_and_profiles(true)
  end

  # You can define as many lanes as you want

  after_all do |lane|

    reset_git_repo

    return unless Helper.is_ci?

    slack(
      message: "Successfully built #{lane}"
    )
  end

  error do |lane, exception|
    return unless Helper.is_ci?

    slack(
      message: exception.message,
      success: false
    )
  end
end

# More information about multiple platforms in fastlane: https://github.com/KrauseFx/fastlane/blob/master/docs/Platforms.md
# All available actions: https://github.com/KrauseFx/fastlane/blob/master/docs/Actions.md
```

### Environment

fastlane version (run `fastlane -v`): 1.87.0

Do you use bundler to execute fastlane (i.e. `bundle exec fastlane`)? no

Do you use a Ruby environment manager (e.g. `chruby`, `rbenv`, `rvm`)? rvm

**Comment by nicolasbraun** [comment]
Can you try leaving only appstore in your CERTIFICATE_TYPES array ? 
Since you are running match 3 times I am not sure how the type is stored, maybe it saves the first run type which in your case in `development`
I suggest you run only match for the type needed by your lane (which is what I do and it works well)

**Comment by nicolasbraun** [comment]
Can you also paste the console output?

**Comment by KrauseFx** [comment]
Can you share a screenshot of your Xcode project's code signing settings.

**Comment by TKBurner** [comment]
@jcampbell05 Thanks for reaching out. Your Fastfile has your fastlane version pinned to an earlier build. Can you update that? Also, I'm not quite sure I understand what you are doing with the type variable in match above. Can you summarize a little bit? Thanks!

**Comment by jcampbell05** [comment]
> Can you try leaving only appstore in your CERTIFICATE_TYPES array ? 
> Since you are running match 3 times I am not sure how the type is stored, maybe it saves the first run type which in your case in development
> I suggest you run only match for the type needed by your lane (which is what I do and it works well)

I updated to do this and it didn't have any effect.

> Can you also paste the console output?

I am running thelane and will be uploading it.

> fastlane member
> KrauseFx commented 6 hours ago
> Can you share a screenshot of your Xcode project's code signing settings.

<img width="597" alt="screen shot 2016-05-17 at 23 54 35" src="https://cloud.githubusercontent.com/assets/1484989/15342087/ba47a2cc-1c8a-11e6-8c55-7bcb280a66c7.png">

Default settings - I know it used to be recommended to set these but they are now deprecated https://github.com/fastlane/fastlane/blob/master/fastlane/docs/CodeSigning.md

> @jcampbell05 Thanks for reaching out. Your Fastfile has your fastlane version pinned to an earlier build. Can you update that? Also, I'm not quite sure I understand what you are doing with the type variable in match above. Can you summarize a little bit? Thanks!

It just a loop which runs match so that it downloads the `development`, `adhoc`, `release` certificates for  the iOS, Watch and Watch Extension.

Updated my fastlane.

Does `fastlane_version` change the behaviour of fastlane ?

**Comment by TKBurner** [comment]
Hi @jcampbell05 Thanks for following up. Felix mentioned this on another thread but we now inherit the settings from Xcode. Are all of your ipa's coming out with your Developer profile?

Thanks!

**Comment by jcampbell05** [comment]
@TKBurner  yes

**Comment by TKBurner** [comment]
@jcampbell05 Thanks for following up. This is now expected behavior. Would it be okay for me to turn this to a feature request so we can use it to gauge interest in reverting to the old behavior?

Thanks!

**Comment by jcampbell05** [comment]
So two things:

1) How do I get the IPA to use the production certificate when built via gym
2) and a feature request to revert or to update the documentation to point out the recommended way to sign the IPA using match so I can push to iTunes via fastlane

**Comment by jcampbell05** [comment]
Ah I think I got it, I need to make my settings be using a team profile.

An updated doc to help people out would be super helpful 😀

**Comment by jcampbell05** [comment]
Ah seems to be bug in a tool I'm using, the behaviour is fine :) in the future I may be building a linter which would guard against these issues.

closing for now.

**Comment by TKBurner** [comment]
@jcampbell05 Awesome! Let me know if you notice anything unexpected with the new behavior.

Hope you have a great week! 👍

**Comment by jcampbell05** [comment]
@TKBurner  I'm reopening as I got this from Apple as I asked them why Xcode defaults to the debug provisioning profiles. 

When using Pilot or Deliver to push to iTunes it would be nice if they could use the Distribution certificate from Match to re-sign the app.

<img width="919" alt="screen shot 2016-06-01 at 09 56 58" src="https://cloud.githubusercontent.com/assets/1484989/15703891/b3b49f86-27df-11e6-95ab-d56e29966dbd.png">

**Comment by KrauseFx** [comment]
@jcampbell05 Interesting, it makes sense: Xcode resigns when exporting your binary. You can also resign using `sigh resign`, however in general there are less problems when setting the correct code signing settings right in Xcode

**Comment by jcampbell05** [comment]
Cool so `sigh resign` after `gym` and before `deliver` ?

**Comment by TKBurner** [comment]
Hi @jcampbell05! Always great to hear from you. Yes, I think that's right, have you tried it? Thanks!

**Comment by jcampbell05** [comment]
@TKBurner  tried it still tricky to use.

How do I get it to use Match and also work with an embeded watch app ?

**Comment by TKBurner** [comment]
@jcampbell05 What specific error are you running into? I think in this case you are going to have to have the provisioning profile set in Xcode rather than set through these steps. Thanks!
